backend/src/services/auth.rs

```diff
@@ -0,0 +1,170 @@
+use crate::router::ApiV1;
+use crate::router::ApiVersion;
+use crate::{
+    connector::BobClient,
+    models::shared::{BobConnectionData, Hostname, RequestTimeout},
+    prelude::*,
+};
+use axum::{
+    async_trait,
+    extract::State,
+    headers::{authorization::Basic, Authorization},
+    Extension, Json,
+};
+use axum_login::{
+    extractors::AuthContext, memory_store::MemoryStore, secrecy::SecretVec, AuthUser, UserStore,
+};
+use hyper::StatusCode;
+use serde::Deserialize;
+use std::{collections::HashMap, fmt::Display, hash::Hash, sync::Arc};
+use thiserror::Error;
+use tokio::sync::RwLock;
+
+pub type SocketBobMemoryStore = MemoryStore<Hostname, BobClient>;
+pub type BobStore<Id> = HashMap<Id, BobClient>;
+pub type SharedBobStore<Id> = Arc<RwLock<BobStore<Id>>>;
+
+#[derive(Debug, Error)]
+pub enum AuthError {
+    #[error("couldn't load user")]
+    LoadError,
+    #[error("couldn't lock user store")]
+    PoisonError,
+}
+
+/// Optional credentials for a BOB cluster
+#[derive(Clone, Debug, PartialEq, Eq, PartialOrd, Ord, Hash, Default, Deserialize)]
+pub struct Credentials {
+    pub login: String,
+    pub password: String,
+}
+
+/// Data needed to login to a BOB cluster
+impl AuthUser<Hostname> for BobClient {
+    fn get_id(&self) -> Hostname {
+        self.hostname().clone()
+    }
+
+    fn get_password_hash(&self) -> SecretVec<u8> {
+        self.context()
+            .get::<Option<Authorization<Basic>>, _>()
+            .as_ref()
+            .map_or_else(
+                || SecretVec::new("".into()),
+                |cred: &Authorization<Basic>| SecretVec::new(cred.password().as_bytes().to_vec()),
+            )
+    }
+}
+
+/// Login to a BOB cluster
+///
+/// # Errors
+/// This function can return the following errors
+///
+/// 1. [`StatusCode::BAD_REQUEST`]
+/// The function failed to parse hostname of the request
+///
+/// 2. [`StatusCode::NOT_FOUND`]
+/// The client was unable to reach the host
+///
+/// 3. [`StatusCode::UNAUTHORIZED`]
+/// The client couldn't authorize on the host
+///
+#[cfg_attr(feature = "swagger", 
+    utoipa::path(
+        post,
+        context_path = ApiV1::to_path(),
+        path = "/login",
+        params(
+            BobConnectionData
+        ),
+        responses(
+            (status = 200, description = "Successful authorization"),
+            (status = 400, description = "Bad Hostname"),
+            (status = 401, description = "Bad Credentials"),
+            (status = 404, description = "Can't reach specified hostname")
+        )
+))]
+pub async fn login(
+    State(store): State<Arc<RwLock<BobStore<Hostname>>>>,
```

Comment:
You have `SharedBobStore<>` type that can be used here instead of full chain of generic types